cranelift/isle/isle/src/printer.rs

```diff
@@ -0,0 +1,250 @@
+//! Printer for ISLE language.
+
+use crate::ast::*;
+use crate::error::Errors;
+use pretty::{Doc, Pretty, RcAllocator, RcDoc};
+use std::io::Write;
+
+/// Printable is a trait satisfied by AST nodes that can be printed.
+pub trait Printable {
+    /// Map the node to a pretty document.
+    fn to_doc(&self) -> RcDoc<()>;
+}
+
+/// Print the given AST node with specified line width.
+pub fn print<N, W>(node: &N, width: usize, out: &mut W) -> Result<(), Errors>
+where
+    N: Printable,
+    W: ?Sized + Write,
+{
+    node.to_doc()
+        .render(width, out)
+        .map_err(|e| Errors::from_io(e, "failed to print isle"))
+}
+
+/// Dump AST node to standard output.
+pub fn dump<N: Printable>(node: &N) -> Result<(), Errors> {
```

Comment:
If possible, I'd rather not have an explicit "... to stdout" in the API -- that's something that the caller can easily wire up, and there's otherwise no particular reason to hardcode one special `Writer` in an API like this. (In general, bad form to reach for "ambient capabilities" rather than those that are passed in, in core libraries, IMHO.)